### ISSUE TITLE
Migration to Gradle 7.3+, for compatibility with the JDK 17 in Android Studio Flamingo

### ISSUE THREAD

**Issue by gnprice** [flutter/flutter#122376]
Gradle (the build system used for Android apps) has an awkwardly strict dependency on the version of Java it runs on: if you try to run a moderately old Gradle on a new JDK, the build will fail. Moreover it will do so with a totally inscrutable error.

This is about to become a problem for lots of Flutter developers, because the Gradle version that `flutter create` would set up until recently is incompatible with the JDK version that lots of developers are going to be upgraded to in a couple of months. Many of them will not realize their JDK was upgraded. Because the Gradle error message is so bad, they will have a frustrating day of debugging, of StackOverflow questions, and likely of taking part in a flood of reports and comments in the Flutter issue tracker.

I think the only feasible way to avoid that flood of frustration is probably to write a migration (like [these](https://github.com/flutter/flutter/tree/main/packages/flutter_tools/lib/src/migrations) and [these](https://github.com/flutter/flutter/tree/main/packages/flutter_tools/lib/src/ios/migrations)) to upgrade existing projects from the Gradle versions `flutter create`  would have chosen in the past to the version `flutter create` would choose today.

### Problem details

The exact Gradle version to use is encoded in a project's tree, at `android/gradle/wrapper/gradle-wrapper.properties`.

Up until #99723, less than a year ago, the default from `flutter create` was Gradle 6.7. So any project created before then is using that version or older, unless it's had an upgrade to the Gradle version.

Starting with Android Studio 2022.2 "Flamingo", the JDK bundled with Android Studio will switch from JDK 11 to JDK 17. That Android Studio version is currently in beta. It will presumably become the new stable in approximately May — at which point developers running existing Android Studio stable versions will start getting prompts to upgrade to it, and large numbers of them will do so.

(Compare the sudden flood of comments on #106674, the last major issue triggered by an Android Studio upgrade, [starting on Jan 12](https://github.com/flutter/flutter/issues/106674). That [was the release date](https://plugins.jetbrains.com/docs/intellij/android-studio-releases-list.html#2022) of 2022.1 "Electric Eel" to stable.)

Whatever JDK comes bundled with Android Studio becomes the JDK that the `flutter` tool uses by default. In doing so, the tool is faithfully following [the recommendation from Android upstream](https://developer.android.com/studio/intro/studio-config#jdk). Many developers, following that same advice, will not even have any other JDK installation lying around.

And Gradle 6.7 is [incompatible with JDK 17](https://docs.gradle.org/current/userguide/compatibility.html). So is any other version before Gradle 7.3. If you try to run an older Gradle with that newer JDK, you get an error like this:

```
FAILURE: Build failed with an exception.

* Where:
Build file '…/example/android/build.gradle'

* What went wrong:
Could not compile build file '…/example/android/build.gradle'.
> startup failed:
  General error during conversion: Unsupported class file major version 61

  java.lang.IllegalArgumentException: Unsupported class file major version 61
  	at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:189)
  	at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:170)
  	[…
  	 …
  	 … 209 more lines of Groovy and Gradle stack trace …
  	 …
  	 …]
  	at java.base/java.lang.Thread.run(Thread.java:833)
```

Putting it together: anyone who created a project with `flutter create` from a Flutter older than last May's Flutter 3.0, and hasn't manually upgraded the version of Gradle it uses, will get the inscrutable error above the next time they attempt `flutter run` or `flutter build` for Android after they acquiesce to Android Studio's next round of self-upgrade prompts in a couple of months.

### Alternate solutions

There are a couple of things we could do to help people choose a JDK independent of their upgrading Android Studio:
* #121501
* #106416

I think we ideally should do those things, on their own merits, but I don't think they really solve this problem. At best, they can only help when the developer knows that what they need is to use another JDK. Because that Gradle error message is so bad, and because the JDK upgrade will have happened as a side effect of upgrading Android Studio, that's already not easy to figure out.

### Workarounds

The existing workarounds I know of are:
* Upgrade Gradle in your project. This is by far the preferred option. Once you've reached the point that you know this is what you need, and once you've worked out the steps involved in an upgrade, it's generally a pretty small and painless change.
* Don't upgrade Android Studio past 2022.1 "Electric Eel".
* Keep an Android Studio "Electric Eel" installation (or older) lying around, even while using a newer version. Then use `flutter config --android-studio-dir=…` to tell Flutter to look at that Android Studio installation, instead of the newer one, when looking for a JDK.

### Why isn't this a problem for Android in general?

If you open an Android project in Android Studio, you get regularly prompted to upgrade Gradle and the Android Gradle Plugin, any time your project isn't already using the latest stable versions. Android Studio provides a slick UI for doing so, too. So anyone working directly in Android will have had ample notice and opportunity to take this upgrade already.

If you open a Flutter project in Android Studio, however, there's no such prompt, even when the project has an `android/` folder with an Android build. It's only if you open that `android/` folder as its own project that you get Android-specific prompts, including those for the Gradle upgrade. If you're happily writing pure Dart code, and not regularly digging into the Android platform side, you may never have cause to see them.

### Other related issues

* Most of the example apps in flutter/packages need this upgrade: #122213 

  (So do apps in `dev/integration_tests/` in flutter/flutter, and apps in the flutter/platform_tests repo.)

## Steps to Reproduce

See the repro steps in #121501. For this issue, only steps 1-4 are needed.

For step 3, instead of going to an existing Flutter project that's using an older Gradle version, one can make the point sharper (for this issue) by running `flutter create` using Flutter 2.10.5 or older. See https://github.com/flutter/flutter/issues/121501#issuecomment-1447718241 .

**Comment by Rexios80** [comment]
I just want to comment on a few things here

- I am using the latest Android Studio Canary, Flutter version 3.7.7, and my gradle versions are up to date
- I still get `Unsupported class file major version` errors _very frequently_ (probably caused bypubspec.yaml changes), and the only way I've found to fix it is to open Android Studio and click the gradle sync button. The flutter tool should probably be doing whatever that button does automatically.
- I still get `Unable to find bundled Java version.` when running `flutter doctor` (kind of unrelated but you did mention that issue)
- Opening the `android/build.gradle` file in Android Stuido is the _only correct way_ to open a Flutter project in Android Studio if you're dealing with native Android code

**Comment by gnprice** [comment]
> * I still get `Unsupported class file major version` errors _very frequently_ (probably caused bypubspec.yaml changes), and the only way I've found to fix it is to open Android Studio and click the gradle sync button. The flutter tool should probably be doing whatever that button does automatically.

Interesting. If you're seeing that when you have a current Gradle version, that would be good to file as a new issue so it can be investigated.


> * I still get `Unable to find bundled Java version.` when running `flutter doctor` (kind of unrelated but you did mention that issue)

This would also be good to report as its own issue. Though take a look first at existing open issues:
https://github.com/flutter/flutter/issues?q=is%3Aopen+label%3Atool+label%3Aplatform-android+unable+to+find+bundled+jdk
to see if one of them covers it.


> * Opening the `android/build.gradle` file in Android Stuido is the _only correct way_ to open a Flutter project in Android Studio if you're dealing with native Android code

If you're dealing with native Android code, then yes, you're going to want to open the `android/` subfolder as its own project in Android Studio. (Though not necessarily the `android/build.gradle` file in particular, if you're just doing work in other files.) When you do, you'll see Android Studio's prompts to upgrade Gradle.

As I said in the OP, though, if you're happily writing pure Dart code, and *not* regularly digging into the Android platform side, then you may never have cause to see those prompts, precisely because you're not opening the `android/` subfolder as its own project. And I think "happily writing pure Dart code" is the state most Flutter developers, most of the time, want to be in and a state our tools aim to support.

**Comment by christopherfujino** [comment]
@reidbaker WDYT about the priority for this?

**Comment by reidbaker** [comment]
@christopherfujino extremely high priority. I am dropping my other work to dig into this. I expect Flamingo to be stable in mid april prior to Google io.

**Comment by christopherfujino** [comment]
Might be related, I tried to regenerate all the gradle lockfiles from our test apps, and I ran into CI failures in https://github.com/flutter/flutter/pull/119290/files

**Comment by reidbaker** [comment]
Update from today's coding session and a conversation with the tools team. 
I have some code that [gets the gradle version](https://github.com/reidbaker/flutter-work/blob/i122376-warn-on-gradle-java-AGP-incompatiblilty/packages/flutter_tools/lib/src/android/gradle_utils.dart#L155), We already [have code that attempts to find the java jdk](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/android/android_sdk.dart#L413) but it does not parse the version. 
I also found code that [parses the AGP version](https://github.com/reidbaker/flutter-work/blob/i122376-warn-on-gradle-java-AGP-incompatiblilty/packages/flutter_tools/lib/src/android/gradle_utils.dart#L146). We also have code that attempts to [match gradle and AGP](https://github.com/reidbaker/flutter-work/blob/i122376-warn-on-gradle-java-AGP-incompatiblilty/packages/flutter_tools/lib/src/android/gradle_utils.dart#L223) 

Unfortunately `flutter doctor` is the wrong context to expose this information since is focused on machine global setup not per project. Apparently `flutter analyze --suggestions` was intended to be "flutter doctor but for projects". 

I do not think this is discoverable enough to protect our users from a confusing error message but after talking to the tools team this is a safe and reasonable place to land code to detect the situation we are worried about. Then we can use/share this code in other parts of the tooling. 

Finally I plan to create a google doc that covers the entire problem of "android studio, java sdk, AGP, Gradle versions and all the ways we are going to try to mitigate the pain for flutter devs including comms" When I do I will post a link here.

**Comment by reidbaker** [comment]
Here is my overarching doc covering this problem since I think it is likely we will have a few bugs to fix and some features to build to mitigate this. https://docs.google.com/document/d/1hTXkjbUrBnXgu8NQsth1c3aEqo77rWoEj8CcsQ39wwQ/edit#

**Comment by reidbaker** [comment]
I am going to try to update this at least 3x a week until I am comfortable with the state. 

I have code that [gets AGP versions from build.gradle](https://github.com/reidbaker/flutter-work/blob/i122376-warn-on-gradle-java-AGP-incompatiblilty/packages/flutter_tools/lib/src/android/gradle_utils.dart#L255) and tests for a new function that [validates agp and gradle](https://github.com/reidbaker/flutter-work/blob/i122376-warn-on-gradle-java-AGP-incompatiblilty/packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart#L327).  Along with improving the documentation from the code written yesterday. 

Next steps are to figure out how the tools team wants me to incorporate the version of java into [project.dart](https://github.com/reidbaker/flutter-work/blob/i122376-warn-on-gradle-java-AGP-incompatiblilty/packages/flutter_tools/lib/src/project.dart#L567). 

Given https://github.com/flutter/flutter/issues/122609#issuecomment-1474293403 I am tempted to write new code that gets the java version and attempts to parse gradle.properties since that will be how gradle decides then mirror the java version code searching done in flutter doctor.

**Comment by reidbaker** [comment]
Update: responded to doc feedback. 

Currently stuck on the design for what code should get the java version. 

We have some code that already finds java in android_sdk.dart
My question (most likely from tools contributors), What is the right way to run code from android_sdk.dart in project.dart? 
My reading shows them as different use-cases (maybe also lifecycles) 
Should I 
A) Have code in project.dart that gets the java version number independent of the code that finds the java binary? 
B) Extract android_sdk code into a static helper and try to give the dependencies from both android_sdk and project.dart 
C) Yall tell me some way those classes are already related and I can use an instance of android_sdk in project.dart (and maybe i feel a bit dumb)
D) something else

Right now I am leaning towards writing new code to find the java version and then dealing with the duplication in code review.

**Comment by gnprice** [comment]
> My question (most likely from tools contributors), What is the right way to run code from android_sdk.dart in project.dart?
> My reading shows them as different use-cases (maybe also lifecycles)

I don't own this code, but FWIW what it looks like to me:

I also read them as relating to different lifecycles or scopes, in that `project.dart` is about an individual Flutter project while `android_sdk.dart` is cross-project and meant to behave the same across whatever Flutter project (or none) you might currently be working on.

But one of those lifecycles is nested inside the other: when you're working on an individual Flutter project, you're going to be using not only some information that's specific to that project but also some of the information that applies across all projects. So while it wouldn't make sense to have `android_sdk.dart` import from `project.dart`, it's perfectly fine and reasonable to go in the reverse direction.

In fact, looking at `project.dart`, it has an import of `gradle_utils.dart`; and that in turn imports `android_sdk.dart`. So there's already a transitive import in `project.dart` of `android_sdk.dart`.

**Comment by gnprice** [comment]
> Unfortunately even though I have android studio installed (an internal version used at google) flutter is not using android studios version of java

One possible cause would be #121925. If your internal version is at least as new/bleeding-edge as the current public canary ("Giraffe"), and given that you're on macOS, then that seems likely.

With the diagnosis in https://github.com/flutter/flutter/issues/121925#issuecomment-1459501923 , it should be possible to tell if your internal version is triggering the same issue.

**Comment by reidbaker** [comment]
Dropping the priority of this bug to correctly reflect the status of the Migrate command as lower than https://github.com/flutter/flutter/issues/123917

**Comment by reidbaker** [comment]
I have code that gets the java version, parses it and that code is not executing as a part of `flutter analyze --suggestitions`. 
https://github.com/flutter/flutter/pull/123916/

Updates will move to the doc or the open pr's.

**Comment by gnprice** [comment]
From some examples @reidbaker pointed out in [the doc](https://github.com/flutter/flutter/issues/122376#issuecomment-1483080507), I'm convinced that an auto-migration for this should stay fairly minimal about when it applies. That's because a Gradle upgrade:
* can introduce warnings, which many projects treat as breaking the build;
* can IIUC sometimes cause errors — potentially that's always due to something misusing Gradle APIs in some way, but
* the error messages can be unhelpful for tracking down what code is at fault (no surprise, given how awful the Gradle error message at the root of this issue thread is); and
* the error might in any case lie not in the app's own code, but in some dependency.

Probably this means the tool should only automatically migrate to a newer Gradle when it's highly confident that not doing so would lead to a broken build, like because it can see the JDK version is a newer one that the existing Gradle version doesn't support. That's the strategy taken in #124085.

**Comment by exaby73** [comment]
In a new project currently, I see the gradle version set it 7.5:

```env
distributionUrl=https\://services.gradle.org/distributions/gradle-7.5-all.zip
```

Can this be considered fixed and hence, can we close this issue?

**Comment by gnprice** [comment]
New projects get a recent version of Gradle, but this issue is about migrating old projects.

I believe #124085 will fix it, though.

**Comment by exaby73** [comment]
Okay thanks @gnprice for the clarification :) I'll update the labels to reflect the current status of the issue

**Comment by reidbaker** [comment]
Last step on this PR is to get a cherry pick landed in 3.10. After that @gmackall can close this ticket.

**Comment by gmackall** [comment]
The cherry-pick of the migrator for this issue just landed and many of the other issues that were referencing this as a tracker have been resolved, so I'm going to close this as completed!

https://github.com/flutter/flutter/pull/126093#issuecomment-1548679586
